sklearn/utils/_array_api.py

```diff
@@ -139,6 +139,13 @@ def __init__(self, array_namespace):
     def __getattr__(self, name):
         return getattr(self._namespace, name)
 
+    @property
+    def float16(self):
```

Comment:
If we want to support `float16`, then we also need to update `isdtype`:

```patch
diff --git a/sklearn/utils/_array_api.py b/sklearn/utils/_array_api.py
index 89560b9370..3371fa6c92 100644
--- a/sklearn/utils/_array_api.py
+++ b/sklearn/utils/_array_api.py
@@ -98,7 +98,10 @@ def _isdtype_single(dtype, kind, *, xp):
                 for k in ("signed integer", "unsigned integer")
             )
         elif kind == "real floating":
-            return dtype in {xp.float32, xp.float64}
+            floats_dtypes = {xp.float32, xp.float64}
+            if hasattr(xp, "float16"):
+                floats_dtypes.add(xp.float16)
+            return dtype in floats_dtypes
         elif kind == "complex floating":
             # Some name spaces do not have complex, such as cupy.array_api
             # and numpy.array_api
@@ -142,7 +145,7 @@ class _ArrayAPIWrapper:
     @property
     def float16(self):
         if self._namespace.__name__ == "numpy.array_api":
-            return numpy.float16
+            return numpy.dtype(numpy.float16)
         else:
             return self._namespace.float16
 
@@ -171,7 +174,7 @@ class _ArrayAPIWrapper:
         return self._namespace.stack(selected, axis=axis)
 
     def isdtype(self, dtype, kind):
-        return isdtype(dtype, kind, xp=self._namespace)
+        return isdtype(dtype, kind, xp=self)
 
 
 def _check_device_cpu(device):  # noqa
@@ -225,6 +228,7 @@ class _NumPyAPIWrapper:
         "uint16",
         "uint32",
         "uint64",
+        "float16",
         "float32",
         "float64",
         "complex64",
diff --git a/sklearn/utils/tests/test_array_api.py b/sklearn/utils/tests/test_array_api.py
index 77fa20e6d0..b941d207d4 100644
--- a/sklearn/utils/tests/test_array_api.py
+++ b/sklearn/utils/tests/test_array_api.py
@@ -253,6 +253,7 @@ def test_get_namespace_array_api_isdtype(wrapper):
     assert xp.isdtype(xp.float32, xp.float32)
     assert xp.isdtype(xp.float32, "real floating")
     assert xp.isdtype(xp.float64, "real floating")
+    assert xp.isdtype(xp.float16, "real floating")
     assert not xp.isdtype(xp.int32, "real floating")
 
     assert xp.isdtype(xp.bool, "bool")

```

Reply:
I added a variation of this, based on the `supported_float_types()` below. Because some Array API implementations support the additional `flaot16` type